### PR TITLE
SNMP migration from snmp.yml to ConfigDB

### PR DESCRIPTION
**Pull request by tsvanduyn**
Initial version on HLD for migration plan.

### CODE REVIEW

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+Commands:
+  add  Add snmp user
+  del  Add snmp user
+admin@str-s6000-acs-11:~$ 
+```
+
+# Migration Plan
+In order to move from using the snmp.yml file to using the Redis ConfigDB here are the things the need to be done to move this in a way that is backward compatible. 
+1. Create a python conversion script to take input from snmp.yml file and convert it to Redis ConfigDB format from above schema.
```
Comment (jleveque) [inline doc/snmp/snmp-configdb-migration-hld.md:208]: For completeness/clarity, this should read,

```
Create a python conversion script to parse the data in snmp.yml file and store it in  ConfigDB using the above schema.
```

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+2. Update Dockerfile.j2 in docker-snmp container to add a line to copy over new python conversion script to "/usr/bin/"
+3. Update the snmpd.conf.j2 jinja template to pull SNMP information from only the Redis ConfigDB
+4. Update start.sh in docker-snmp container to add a line above sonic-cfggen to run python conversion script (which will run everytime this docker container starts) and then comment out the "-y /etc/sonic/snmp.yml" file as this is not needed since we grabbed the snmp.yml info in the python conversion script and have the information available in the Redis ConfigDB.
```
Comment (jleveque) [inline doc/snmp/snmp-configdb-migration-hld.md:211]: Suggest rewording as follows:

```
4. Update start.sh in docker-snmp container as follows:
    - Add a line above the `sonic-cfggen` call to run the python conversion script from step 1 (which will run every time this docker container starts)
    - Remove the `-y /etc/sonic/snmp.yml` argument from the existing call to sonic-cfggen as it will no longer needed because the python conversion script above will have already loaded that data into Config DB
```

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+admin@str-s6000-acs-11:~$ 
+```
+
+# Migration Plan
+In order to move from using the snmp.yml file to using the Redis ConfigDB here are the things the need to be done to move this in a way that is backward compatible. 
+1. Create a python conversion script to take input from snmp.yml file and convert it to Redis ConfigDB format from above schema.
+2. Update Dockerfile.j2 in docker-snmp container to add a line to copy over new python conversion script to "/usr/bin/"
+3. Update the snmpd.conf.j2 jinja template to pull SNMP information from only the Redis ConfigDB
+4. Update start.sh in docker-snmp container to add a line above sonic-cfggen to run python conversion script (which will run everytime this docker container starts) and then comment out the "-y /etc/sonic/snmp.yml" file as this is not needed since we grabbed the snmp.yml info in the python conversion script and have the information available in the Redis ConfigDB. 
+5. Create new docker-snmp container with all these updates so that we will eventually be able to remove the snmp.yml file and only use the Redis ConfigDB after we socalize the update.
```
Comment (jleveque) [inline doc/snmp/snmp-configdb-migration-hld.md:212]: This step seems unnecessary, as once you make the above changes, every time the container is built it will have the changes above. Maybe just add a note or a new section at the end of the file about the steps necessary to completely remove support for the snmp.yml file when we are ready to do so.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 1 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: updated.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 2 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: Updated.  I like your wording here a lot better.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 3 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: Added a new notes section for this.

**Review by jleveque — APPROVED** [review]
Looks good to me. Please wait for other reviewers.

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+admin@str-s6000-acs-11:~$ show run snmp -h
+Usage: show run snmp [OPTIONS] COMMAND [ARGS]...
+
+  Show SNMP information
+
+Options:
+  -?, -h, --help  Show this message and exit.
+
+Commands:
+  community  show runningconfiguration snmp community
```
Comment (SuvarnaMeenakshi) [inline doc/snmp/snmp-configdb-migration-hld.md:85]: space between "runningconfiguration"

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+sudo config snmp community 
+```
+admin@str-s6000-acs-11:~$ sudo config snmp community -h
+Usage: config snmp community [OPTIONS] COMMAND [ARGS]...
+
+Options:
+  -?, -h, --help  Show this message and exit.
+
+Commands:
+  add  Add snmp community
```
Comment (SuvarnaMeenakshi) [inline doc/snmp/snmp-configdb-migration-hld.md:161]: how will the type of the community be added using this command?

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  add  Add snmp user
+  del  Add snmp user
+admin@str-s6000-acs-11:~$ 
+```
+
+# Migration Plan
+In order to move from using the snmp.yml file to using the Redis ConfigDB here are the things the need to be done to move this in a way that is backward compatible. 
+1. Create a python conversion script to parse the data in snmp.yml file and store it in ConfigDB using the above schema.
+2. Update Dockerfile.j2 in docker-snmp container to add a line to copy over new python conversion script to "/usr/bin/"
+3. Update the snmpd.conf.j2 jinja template to pull SNMP information from only the Redis ConfigDB
```
Comment (SuvarnaMeenakshi) [inline doc/snmp/snmp-configdb-migration-hld.md:210]: Initially, The SNMP information will be pulled from both configDB and snmp.yml, as mentioned in the PR: https://github.com/Azure/sonic-buildimage/pull/6134. So step 3 should be to use both till complete migration is done, is that right?

**Review by SuvarnaMeenakshi — CHANGES_REQUESTED** [review]

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 4 above)
Comment (jleveque) [reply doc/snmp/snmp-configdb-migration-hld.md:85]: In the SONiC CLI, `runningconfiguration` (no spaces) is a subcommand of `show`

Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-configdb-migration-hld.md:85]: As this is a help string, looking at the help strings of the existing commands, I thought it would be good to keep it uniform.
 show running --help
Usage: show running [OPTIONS] COMMAND [ARGS]...
  Show current running configuration information
Commands:
  acl         Show acl running configuration

Comment (jleveque) [reply doc/snmp/snmp-configdb-migration-hld.md:85]: Ah, I understand your point, now. The help string is currently just the command syntax. It should be a more human-readable description of what the command does.

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+SNMP|LOCATION
+SNMP|CONTACT
+
+admin@str-s6000-acs-11:~$ redis-cli -n 4 hgetall "SNMP|LOCATION"
+1) "LOCATION"
+2) "Redmond"
+
+admin@str-s6000-acs-11:~$ redis-cli -n 4 hgetall "SNMP|CONTACT"
+1) "joe"
+2) "joe@microsoft.com"
```
Comment (qiluo-msft) [inline doc/snmp/snmp-configdb-migration-hld.md:39]: Use a demo email address like `joe@contoso.com`

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+# SNMP Table
+```
+SNMP|LOCATION
+SNMP|CONTACT
+
+admin@str-s6000-acs-11:~$ redis-cli -n 4 hgetall "SNMP|LOCATION"
+1) "LOCATION"
+2) "Redmond"
+
+admin@str-s6000-acs-11:~$ redis-cli -n 4 hgetall "SNMP|CONTACT"
```
Comment (qiluo-msft) [inline doc/snmp/snmp-configdb-migration-hld.md:37]: Use a demo machine name like `switch1`

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+
+# SNMP Table
+```
+SNMP|LOCATION
+SNMP|CONTACT
+
+admin@str-s6000-acs-11:~$ redis-cli -n 4 hgetall "SNMP|LOCATION"
+1) "LOCATION"
+2) "Redmond"
```
Comment (qiluo-msft) [inline doc/snmp/snmp-configdb-migration-hld.md:35]: Use a fictional location like `Emerald City`

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+Commands:
+  add  Add snmp user
+  del  Add snmp user
+admin@str-s6000-acs-11:~$ 
+```
+
+# Migration Plan
+In order to move from using the snmp.yml file to using the Redis ConfigDB here are the things the need to be done to move this in a way that is backward compatible. 
+1. Create a python conversion script to parse the data in snmp.yml file and store it in ConfigDB using the above schema.
```
Comment (qiluo-msft) [inline doc/snmp/snmp-configdb-migration-hld.md:208]: The SNMP information will be pulled from both configDB and snmp.yml. Why you still need a conversion script?

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+```
+admin@str-s6000-acs-11:~$ sudo config snmp community -h
+Usage: config snmp community [OPTIONS] COMMAND [ARGS]...
+
+Options:
+  -?, -h, --help  Show this message and exit.
+
+Commands:
+  add  Add snmp community
+  del  Add snmp community
```
Comment (qiluo-msft) [inline doc/snmp/snmp-configdb-migration-hld.md:162]: Instead of `add` `del` individual community string, better to `set` multiple strings in one command. So there will never be a transient state, such as no string at all.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 5 above)
Comment (qiluo-msft) [reply doc/snmp/snmp-configdb-migration-hld.md:161]: Agree, this command line syntax is not fully defined. Many more below.

**Review by qiluo-msft — CHANGES_REQUESTED** [review]

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 4 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: I updated this document.  This information will be more flushed out during the PR's to submit the commands once we've finalized the details on the design.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 5 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md:161]: Per my conversation with Suvarna I added a section with with how it would look for showing the full output.  
This was already setup but I didn't put it in the doc as I thought it would be more relevant to the PR when the commands were submitted.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 6 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md:245]: Per my conversation with Suvarna with this setup we will just be able to pull the snmp.yml info into configdb with the conversion script when the docker container starts.  This way we will always get the info from the snmp.yml file but allow us to use the configdb.   This way the jinja template only has to support the configdb since the snmp.yml file info was already imported we will get both.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 7 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: fixed.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 8 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: fixed.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 9 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: fixed.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 10 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md:243]: If we pull the info from the snmp.yml file into the configdb whenever the docker-snmp container is restarted it makes it much cleaner for show/config commands and allows for backward compatibility.  This will be much cleaner.  I've discussed this with Suvarna and Joe but didn't get a chance to run it by you yet.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 11 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: So the none of the other networking vendor OS allow this.  All vendors only support adding and removing one at a time. I think we should follow that standard.   We should be able to have no string all if if that is what folks want.

File: doc/snmp/snmp-configdb-migration-hld.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+```
+admin@switch1:~$ sudo config snmp community -h
+Usage: config snmp community [OPTIONS] COMMAND [ARGS]...
+
+Options:
+  -?, -h, --help  Show this message and exit.
+
+Commands:
+  add  Add snmp community
+  del  Delete snmp community
```
Comment (SuvarnaMeenakshi) [inline doc/snmp/snmp-configdb-migration-hld.md:162]: Minor suggestion, might be good to have "Add snmp community string"

**Review by SuvarnaMeenakshi — APPROVED** [review]
LGTM.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 12 above)
Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md:162]: will do this in the PR for the commands.

File: doc/snmp/snmp-configdb-migration-hld.md (hunk 11 above)
Comment (qiluo-msft) [reply doc/snmp/snmp-configdb-migration-hld.md]: Suggest keep `add` `del` commands, and add a `replace` command, which is useful in your use case.

---
In reply to: [541519000](https://github.com/Azure/SONiC/pull/723#discussion_r541519000) [](ancestors = 541519000)

Comment (tsvanduyn) [reply doc/snmp/snmp-configdb-migration-hld.md]: Added replace command.

**Review by qiluo-msft — APPROVED** [review]
